src/agent/thread_helper.cpp

```diff
@@ -399,6 +411,82 @@ void ThreadHelper::LogOpenThreadResult(const char *aAction, otError aError)
     }
 }
 
+void ThreadHelper::AttachToNetwork(const std::vector<uint8_t> &aDatasetTlvs, ThreadHelper::ResultHandler aHandler)
+{
+    otError                  error = OT_ERROR_NONE;
+    otOperationalDatasetTlvs datasetTlvs;
+    otOperationalDataset     existingDataset;
+    otDeviceRole             role = otThreadGetDeviceRole(mInstance);
+
+    VerifyOrExit(aHandler != nullptr, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(mAttachHandler == nullptr && mJoinerHandler == nullptr, error = OT_ERROR_INVALID_STATE);
+
+    VerifyOrExit(aDatasetTlvs.size() <= sizeof(datasetTlvs.mTlvs), error = OT_ERROR_INVALID_ARGS);
+    std::copy(aDatasetTlvs.begin(), aDatasetTlvs.end(), datasetTlvs.mTlvs);
+    datasetTlvs.mLength = aDatasetTlvs.size();
+
+    mAttachHandler = aHandler;
+
+    if (role == OT_DEVICE_ROLE_DISABLED || role == OT_DEVICE_ROLE_DETACHED)
+    {
+        error = otDatasetGetActive(mInstance, &existingDataset);
+        VerifyOrExit(error == OT_ERROR_NONE || error == OT_ERROR_NOT_FOUND);
+
+        if (error == OT_ERROR_NOT_FOUND)
+        {
+            SuccessOrExit(error = otDatasetSetActiveTlvs(mInstance, &datasetTlvs));
+
+            if (!otIp6IsEnabled(mInstance))
+            {
+                SuccessOrExit(error = otIp6SetEnabled(mInstance, true));
+            }
+            SuccessOrExit(error = otThreadSetEnabled(mInstance, true));
+
+            ExitNow();
+        }
+
+        ExitNow(error = OT_ERROR_INVALID_STATE);
+    }
+
+    SuccessOrExit(error = otDatasetParseTlvs(&datasetTlvs, &mAttachingDataset));
+    SuccessOrExit(error = otDatasetGetActive(mInstance, &existingDataset));
+
+    if (IsSameDataset(mAttachingDataset, existingDataset))
+    {
+        ExitNow(error = OT_ERROR_ALREADY);
+    }
+
+    SuccessOrExit(error = otDatasetSendMgmtPendingSet(mInstance, &mAttachingDataset, nullptr, 0));
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        if (aHandler)
+        {
+            aHandler(error);
+        }
+        mAttachHandler = nullptr;
```

Comment:
Are there cases that we will overwrite the handler set by a previously successful call of `AttachToNetwork`?

Reply:
This is indeed an issue, and it also happens in `Attach` and `JoinerStart`. Any idea how to solve it elegantly?

Reply:
Resolved by moving all assignments to it to the end of the function.